src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "AtlasEngine.h"
+
+// #### NOTE ####
+// If you see any code in here that contains "_r." you might be seeing a race condition.
+// The AtlasEngine::Present() method is called on a background thread without any locks,
+// while any of the API methods (like AtlasEngine::Invalidate) might be called concurrently.
+// The usage of _r fields is unsafe as those are accessed and written to by the Present() method.
+
+#pragma warning(disable : 4100) // '...': unreferenced formal parameter
+// Disable a bunch of warnings which get in the way of writing performant code.
+#pragma warning(disable : 26429) // Symbol 'data' is never tested for nullness, it can be marked as not_null (f.23).
+#pragma warning(disable : 26446) // Prefer to use gsl::at() instead of unchecked subscript operator (bounds.4).
+#pragma warning(disable : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+#pragma warning(disable : 26482) // Only index into arrays using constant expressions (bounds.2).
+
+using namespace Microsoft::Console::Render;
+
+// Like gsl::narrow but returns a HRESULT.
+#pragma warning(push)
+#pragma warning(disable : 26472) // Don't use a static_cast for arithmetic conversions. Use brace initialization, gsl::narrow_cast or gsl::narrow (type.1).
+template<typename T, typename U>
+constexpr HRESULT api_narrow(U val, T& out) noexcept
+{
+    out = static_cast<T>(val);
+    return static_cast<U>(out) != val || (std::is_signed_v<T> != std::is_signed_v<U> && out < T{} != val < U{}) ? HRESULT_FROM_WIN32(ERROR_ARITHMETIC_OVERFLOW) : S_OK;
+}
+#pragma warning(pop)
+
+template<typename T, typename U>
+constexpr HRESULT vec2_narrow(U x, U y, AtlasEngine::vec2<T>& out) noexcept
+{
+    return api_narrow(x, out.x) | api_narrow(y, out.y);
+}
+
+#pragma region IRenderEngine
+
+[[nodiscard]] HRESULT AtlasEngine::Invalidate(const SMALL_RECT* const psrRegion) noexcept
+{
+    //assert(psrRegion->Top < psrRegion->Bottom && psrRegion->Top >= 0 && psrRegion->Bottom <= _api.cellCount.y);

```

Comment:
>//assert(ps [](http://example.com/codeflow?start=4&length=11)

Why leave these commented here and below? Don't they get compiled out in release anyway?

Reply:
A number of actions with conhost (like resizing, font changes, or simply pausing the application with a debugger during startup, so that a race condition between conhost's main thread and the kernel sending the first WM_SIZE message can take hold) cause invalid coordinates to be passed to our renderers. I was getting a bit annoyed by this and commented out the assertion for now. I'll uncomment them when I'll debug this issue for EIM.